crates/bevy_ecs/src/index/query_by_index.rs

```diff
@@ -0,0 +1,304 @@
+    where
+        QueryState<D, (F, With<C>)>: Clone,
+    {
+        let state = self.state.primary_query_state.clone();
+
+        // SAFETY: We have registered all of the query's world accesses,
+        // so the caller ensures that `world` has permission to access any
+        // world data that the query needs.
+        unsafe {
+            QueryLens::new(
+                self.world,
+                self.filter_for_value(value, state),
+                self.last_run,
+                self.this_run,
+            )
+        }
+    }
+
+    /// Return a read-only [`QueryLens`] returning entities with a component `C` of the provided value.
+    pub fn at(&self, value: &C) -> QueryLens<'_, D::ReadOnly, (F, With<C>)>
+    where
+        QueryState<D::ReadOnly, (F, With<C>)>: Clone,
+    {
+        let state = self.state.primary_query_state.as_readonly().clone();
+
+        // SAFETY: We have registered all of the query's world accesses,
+        // so the caller ensures that `world` has permission to access any
+        // world data that the query needs.
+        unsafe {
+            QueryLens::new(
+                self.world,
+                self.filter_for_value(value, state),
+                self.last_run,
+                self.this_run,
+            )
+        }
+    }
+
+    fn filter_for_value<T: QueryData, U: QueryFilter>(
+        &self,
+        value: &C,
+        mut state: QueryState<T, U>,
+    ) -> QueryState<T, U> {
+        match self.index.mapping.get(value) {
+            Some(index) => {
+                state = (0..self.index.markers.len())
+                    .map(|i| (i, 1 << i))
+                    .take_while(|&(_, mask)| mask <= self.index.slots.len())
+                    .map(|(i, mask)| {
+                        if index & mask > 0 {
+                            &self.state.with_states[i]
+                        } else {
+                            &self.state.without_states[i]
+                        }
+                    })
+                    .fold(state, |state, filter| {
+                        state.join_filtered(self.world, filter)
+                    });
+            }
+            None => {
+                // Create a no-op filter by joining two conflicting filters together.
+                let filter = &self.state.with_states[0];
+                state = state.join_filtered(self.world, filter);
+
+                let filter = &self.state.without_states[0];
+                state = state.join_filtered(self.world, filter);
+            }
+        }
+
+        state
+    }
+}
+
+#[doc(hidden)]
+pub struct QueryByIndexState<
+    C: Component<Mutability = Immutable>,
+    D: QueryData + 'static,
+    F: QueryFilter + 'static,
+> {
+    primary_query_state: QueryState<D, (F, With<C>)>,
+    index_state: ComponentId,
+
+    // TODO: THERE MUST BE A BETTER WAY
+    without_states: Vec<QueryState<(), With<C>>>, // No, With<C> is not a typo
```

Comment:
Comment needs to be clearer here. Why on earth does this work?

Reply:
For what it's worth, I have such a follow up PR mostly ready to go, with the design I mentioned on Discord.  So I wouldn't worry too much how well documented this version is.  And it would be nice not to have too many rebase conflicts :).